src/inference/AdvancedSMC.jl

```diff
@@ -342,17 +342,17 @@ function DynamicPPL.assume(
             r = vi[vn]
         else
             r = rand(rng, dist)
-            push!(vi, vn, r, dist, DynamicPPL.Selector(:invalid))
+            push!!(vi, vn, r, dist, DynamicPPL.Selector(:invalid))
         end
         lp = logpdf_with_trans(dist, r, istrans(vi, vn))
-        acclogp!(vi, lp)
+        acclogp!!(vi, lp)
     end
-    return r, 0
+    return r, 0, _vi
```

Comment:
Is this the right thing to do @yebai @devmotion ?

Essentially the issue is that when we're using `PG` with threads, `_vi` is a `ThreadSafeVarInfo` while `vi = AdvancedPS.current_trace().f.varinfo` is _not_. And so if we return `vi` instead of `_vi`, subsequent code in `evaluate_threadsafe!!` will fail.

I'm not too familiar with the AdvancedPS stuff so it's a bit unclear to me what I should be returning here, e.g. is the above okay?

Reply:
Or should I overload `evaluate!!` and all the corresponding methods?

If we're doing this, we might want to consider making the "outer" calls, e.g. `(::Model)(...)` a bit more general, e.g. dispatch on `AbstractProbabilisticProgram` instead and then _just_ overloading `evaluate_threadsafe!!` and `evaluate_threadunsafe!!` for `TracedModel`.

Reply:
Let's perhaps disallow the use of `ThreadSafeVarInfo` with `PG`/`SMC` for now. These samplers are not yet ready to run in threads parallelism due to the interaction of particles (i.e. the `resampling` step). I suggest that we simply throw an error here explaining that `ThreadSafeVarInfo` support for `PG` is to be supported in the near future.

Reply:
Well, there currently isn't a way to allow/disallow specific combinations like this. And erroring or even just telling people not to use `PG`/`SMC` just because they are running with `--threads=4` seems wrong since they might not even be using threading inside of the model.

We _could_ add a `isthreadsafe(varinfo, context)` check in https://github.com/TuringLang/DynamicPPL.jl/blob/57c50f1f46aa5e51ec2125f3ae4883c61e59e9c3/src/model.jl#L391 and then implement this specifically for `SamplingContext{<:PG}`, etc.

Reply:
> We _could_ add a `isthreadsafe(varinfo, context)` check in https://github.com/TuringLang/DynamicPPL.jl/blob/57c50f1f46aa5e51ec2125f3ae4883c61e59e9c3/src/model.jl#L391 and then implement this specifically for `SamplingContext{<:PG}`, etc.

Yes, I think this would be a better approach. Then we can always choose the non-threadsafe VarInfo for these samplers instead of erroring if they run Julia with multiple threads (e.g. I always do, so I think it would be very inconvenient to throw an error here).

The only possible problem is that then it would be possible to use `@threads` with a non-threadsafe VarInfo which can lead to incorrect result silently. I don't think there's a clean way to detect if a model tries to use multiple threads (since it is not as simple as detecting a `@threads` macro call). I wonder if it is possible to add a locking mechanism to the default non-threadsafe VarInfo that makes it threadsafe while not impacting performance if it is run with a single thread. Surely it would be slower if multiple threads are used but with multiple threads we would use the more performant ThreadSafeVarInfo anyway - if we don't use PG etc. The main motivation would be to avoid silent problems with multithreaded models when PG + a non-threadsafe VarInfo is used - it seems fine if it is less performant since it is not recommended.

Reply:
> Yes, I think this would be a better approach. Then we can always choose the non-threadsafe VarInfo for these samplers instead of erroring if they run Julia with multiple threads (e.g. I always do, so I think it would be very inconvenient to throw an error here).

I'll do that then :+1:

> The only possible problem is that then it would be possible to use @threads with a non-threadsafe VarInfo which can lead to incorrect result silently.

Yup :confused: We could potentially address this by using atomics (though it would probably require very recent Julia versions to be able to do this properly) which I expect would behave just like a `Ref` in the single-threaded case, or even just making the default `VarInfo` threadsafe by using an array instead of `Ref` for the logp field.

But it's worth mentioning that currently `PG`, etc. already silently does the wrong thing, doesn't it?